legacy/firmware/messages.c

```diff
@@ -291,43 +298,56 @@ const uint8_t *msg_out_data(void) {
 
 const uint8_t *msg_debug_out_data(void) {
   if (msg_debug_out_start == msg_debug_out_end) return 0;
-  uint8_t *data = msg_debug_out + (msg_debug_out_start * 64);
-  msg_debug_out_start = (msg_debug_out_start + 1) % (MSG_DEBUG_OUT_SIZE / 64);
+  uint8_t *data = msg_debug_out + (msg_debug_out_start * USB_PACKET_SIZE);
+  msg_debug_out_start =
+      (msg_debug_out_start + 1) % (MSG_DEBUG_OUT_BUFFER_SIZE / USB_PACKET_SIZE);
   debugLog(0, "", "msg_debug_out_data");
   return data;
 }
 
 #endif
 
+// msg_tiny needs to be large enough to hold the C struct decoded from a single
+// 64 byte USB packet. The decoded struct can be larger than the encoded
+// protobuf message. However, 128 bytes should be more than enough.
 CONFIDENTIAL uint8_t msg_tiny[128];
 _Static_assert(sizeof(msg_tiny) >= sizeof(Cancel), "msg_tiny too tiny");
+_Static_assert(USB_PACKET_SIZE >= MSG_HEADER_SIZE + Cancel_size,
+               "msg_tiny too tiny");
 _Static_assert(sizeof(msg_tiny) >= sizeof(Initialize), "msg_tiny too tiny");
+_Static_assert(USB_PACKET_SIZE >= MSG_HEADER_SIZE + Initialize_size,
+               "msg_tiny too tiny");
 _Static_assert(sizeof(msg_tiny) >= sizeof(PassphraseAck), "msg_tiny too tiny");
+_Static_assert(USB_PACKET_SIZE >= MSG_HEADER_SIZE + PassphraseAck_size,
+               "msg_tiny too tiny");
 _Static_assert(sizeof(msg_tiny) >= sizeof(ButtonAck), "msg_tiny too tiny");
+_Static_assert(USB_PACKET_SIZE >= MSG_HEADER_SIZE + ButtonAck_size,
+               "msg_tiny too tiny");
 _Static_assert(sizeof(msg_tiny) >= sizeof(PinMatrixAck), "msg_tiny too tiny");
+_Static_assert(USB_PACKET_SIZE >= MSG_HEADER_SIZE + PinMatrixAck_size,
+               "msg_tiny too tiny");
 #if DEBUG_LINK
 _Static_assert(sizeof(msg_tiny) >= sizeof(DebugLinkDecision),
                "msg_tiny too tiny");
+_Static_assert(USB_PACKET_SIZE >= MSG_HEADER_SIZE + DebugLinkDecision_size,
+               "msg_tiny too tiny");
 _Static_assert(sizeof(msg_tiny) >= sizeof(DebugLinkGetState),
                "msg_tiny too tiny");
+_Static_assert(USB_PACKET_SIZE >= MSG_HEADER_SIZE + DebugLinkGetState_size,
+               "msg_tiny too tiny");
 #endif
 uint16_t msg_tiny_id = 0xFFFF;
 
 void msg_read_tiny(const uint8_t *buf, int len) {
-  if (len != 64) return;
-  if (buf[0] != '?' || buf[1] != '#' || buf[2] != '#') {
+  if (len != USB_PACKET_SIZE || buf[0] != '?' || buf[1] != '#' ||
+      buf[2] != '#') {
+    // Ignore unexpected messages. This is helpful when two applications are
```

Comment:
This is actually ignoring unexpected _packets_; a start-of-message packet will go through even if it is a wrong one.

Reply:
Changed to "Ignore unexpected packets.", https://github.com/trezor/trezor-firmware/pull/1466/commits/b88002bdd809ce29871ec84bc9894bd456c4a1c6#diff-c3507f7877d6bbfc9fc51af19aa85cf572933ecfd56b608b88c6935f0b703a7cR343